src/DataFramesMeta.jl

```diff
@@ -397,33 +397,32 @@ end
 ##
 ##############################################################################
 
-# needed on Julia 1.0 till #1489 in DataFrames is merged
-orderby(d::DataFrame, arg::DataFrame) = d[sortperm(arg), :]
+function orderby_helper(x, args...)
+    t = (fun_to_vec(arg; nolhs = true) for arg in args)
+    quote
+        $DataFramesMeta.orderby($x, $(t...))
+    end
+end
 
-function orderby(d::AbstractDataFrame, args...)
-    D = typeof(d)(args...)
-    d[sortperm(D), :]
+function orderby(x::AbstractDataFrame, @nospecialize(args...))
+    t = DataFrames.select(x, args...; copycols = false)
+    x[sortperm(t), :]
 end
 
-orderby(d::AbstractDataFrame, f::Function) = d[sortperm(f(d)), :]
-orderby(g::GroupedDataFrame, f::Function) = g[sortperm([f(x) for x in g])]
+function orderby(x::GroupedDataFrame, @nospecialize(args...))
 
-orderbyconstructor(d::AbstractDataFrame) = (x...) -> DataFrame(Any[x...], Symbol.(1:length(x)))
-orderbyconstructor(d) = x -> x
+    @warn "orderby behavior now returns a `DataFrame` instead of a `GroupedDataFrame`. " *
+          "Group the returned data frame to restore old behavior" maxlog = 2
 
-function orderby_helper(d, args...)
-    _D = gensym()
-    quote
-        let $_D = $d
-            $orderby($_D, $(with_anonymous(:($orderbyconstructor($_D)($(args...))))))
-        end
-    end
+    t = DataFrames.select(x, args...; copycols = false, keepkeys = false)
+    parent(x)[sortperm(t), :]
```

Comment:
Is this correct if some groups have been dropped?

Reply:
Can you clarify circumstances in which groups can be dropped? That should never happen with `select`, right?

Reply:
in `select` groups are not allowed to be dropped, but can be reordered (but reordering is ignored).

Reply:
But `x` could have been subsetted before calling `orderby`. Isn't that a problem since `sortperm(t)` will return a permutation of rows that have not been dropped? It's easy to check by adding `@assert nrow(t) == nrow(parent(x))`.

Reply:
`select` should fail in that case, but I will add the check to ensure.

Reply:
yes - this is what I meant. `select` will fail if `x` is subsetted, but will not fail if `x` is just reordered